tests/unit-tests/src/VuFindTest/OaiPmh/CommunicatorTest.php

```diff
@@ -0,0 +1,180 @@
+            ->method('isSuccess')
+            ->will($this->returnValue(true));
+        $response->expects($this->exactly(2))
+            ->method('getStatusCode')
+            ->willReturnOnConsecutiveCalls(503, 200);
+        $response->expects($this->any())
+            ->method('getBody')
+            ->will($this->returnValue($this->getIdentifyResponse()));
+        $header = $this->getMockBuilder(\Laminas\Http\Header\RetryAfter::class)
+            ->getMock();
+        $header->expects($this->once())
+            ->method('getDeltaSeconds')
+            ->will($this->returnValue(1));
+        $headers = $response->getHeaders();
+        $headers->expects($this->any())
+            ->method('get')
+            ->with($this->equalTo('Retry-After'))
+            ->will($this->returnValue($header));
+        $uri = 'http://localhost';
+        $comm = $this->getCommunicator($uri, $client);
+        $this->assertEquals(
+            $comm->request('Identify'),
+            $this->getIdentifyResponse()
+        );
+    }
+
+    /**
+     * Test communicator HTTP error detection.
+     *
+     * @return void
+     *
+     */
+    public function testHTTPErrorDetection()
+    {
+        $this->expectException(\Exception::class);
+        $this->expectExceptionMessage('HTTP Error');
+
+        $client = $this->getMockClient();
+        $response = $client->send();
+        $response->expects($this->any())
+            ->method('isSuccess')
+            ->will($this->returnValue(false));
+        $uri = 'http://localhost';
+        $comm = $this->getCommunicator($uri, $client);
+        $comm->request('Identify');
+    }
+
+    // Internal API
+
+    /**
+     * Get a sample Identify response
+     *
+     * @return string
+     */
+    protected function getIdentifyResponse()
+    {
+        return '<?xml version="1.0"?><OAI-PMH xmlns="http://www.openarchives.org/OAI/2.0/" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://www.openarchives.org/OAI/2.0/ http://www.openarchives.org/OAI/2.0/OAI-PMH.xsd"><responseDate>2013-10-11T11:12:04Z</responseDate><request verb="Identify" submit="Go">http://fake/my/OAI/Server</request><Identify><repositoryName>myuniversity University VuFind</repositoryName><baseURL>http://fake/my/OAI/Server</baseURL><protocolVersion>2.0</protocolVersion><earliestDatestamp>2000-01-01T00:00:00Z</earliestDatestamp><deletedRecord>transient</deletedRecord><granularity>YYYY-MM-DDThh:mm:ssZ</granularity><adminEmail>libtech@myuniversity.edu</adminEmail><description><oai-identifier xmlns="http://www.openarchives.org/OAI/2.0/oai-identifier" xsi:schemaLocation="http://www.openarchives.org/OAI/2.0/oai-identifier http://www.openarchives.org/OAI/2.0/oai-identifier.xsd"><scheme>oai</scheme><repositoryIdentifier>fake.myuniversity.edu</repositoryIdentifier><delimiter>:</delimiter><sampleIdentifier>oai:fake.myuniversity.edu:123456</sampleIdentifier></oai-identifier></description></Identify></OAI-PMH>';
```

Comment:
I'm not sure if you really need to use a real response here, since the communicator doesn't care about the contents of the requests it's making. If you want to avoid the very long line, you could do something like `<mock />` and I expect the results would be the same.